HV-976 Optional and wraper

DavideD opened the pull request:
https://hibernate.atlassian.net/browse/HV-976
https://hibernate.atlassian.net/browse/HV-977

@hferentschik 
this solution will pass all the test but I'm not sure it's the correct one.

In particular, what is the expected behaviour for his scenario (does it make sense)?:

```
@UnwrapValidatedValue(true)
@NotNull(message = "container")
Optional<@NotNullTypeUse(message = "type") String> valueWithNotNull;
```

gunnarmorling commented:
> In particular, what is the expected behaviour for his scenario

On first thought I'd say both constraints should be applied to the String value:

* `@NotNull` as unwrap=true is given
* `@NotNullTypeUse` as it is given on the type use of `String`

Is that how it does behave?

DavideD commented:
What would you expect instead if UnwrapValidatedValue is set to false (`@UnwrapValidatedValue(false)`)?

DavideD commented:
I think a test case is missing for that scenario, I will add one.

gunnarmorling commented:
> What would you expect instead if UnwrapValidatedValue is set to false (@UnwrapValidatedValue(false))? 

Then I'd expect `@NotNull` to be applied to the `Optional` and `@NotNullTypeUse` to be applied to the wrapped String.

gunnarmorling commented on engine-jdk8-tests/src/test/java/org/hibernate/validator/test/internal/engine/valuehandling/OptionalTypeAnnotationConstraintTest.java:
```diff
@@ -0,0 +1,216 @@
+import static org.hibernate.validator.testutil.ValidatorUtil.getValidator;
+
+/**
+ * @author Lukas Niemeier
```
Can we include this? Do we have a CLA from Lukas?

DavideD replied:
I don't know, we need to ask him, but these are tests he wrote so I feel bad not to mention him

hferentschik replied:
Since he provided the code on JIRA without any authorship I think it would be safer to remove it. In some form he still has credit via the issue itself. This avoids any problems.

DavideD replied:
Ok, no problem

gunnarmorling commented on engine-jdk8-tests/src/test/java/org/hibernate/validator/test/internal/engine/valuehandling/OptionalTypeAnnotationConstraintTest.java:
```diff
@@ -0,0 +1,216 @@
+		assertNumberOfViolations( constraintViolations, 0 );
+	}
+
+	/**
```
I guess the "Actual..." comments can go away now.

gunnarmorling commented on engine/src/main/java/org/hibernate/validator/internal/metadata/provider/AnnotationMetaDataProvider.java:
```diff
@@ -262,6 +256,29 @@ private ConstrainedField findPropertyMetaData(Field field) {
 		);
 	}
 
+	private UnwrapMode unwrapMode(Field field, boolean typeArgumentAnnotated) {
```
Shouldn't something similar also be done for (getter-defined) properties and method-level constraints?

DavideD replied:
Mmmh , make sense, I will add the tests

gunnarmorling commented on engine/src/main/java/org/hibernate/validator/internal/engine/ValidatorImpl.java:
```diff
@@ -530,11 +532,19 @@ private boolean validateConstraint(ValidationContext<?> validationContext,
 				ReflectionHelper.getPropertyName( metaConstraint.getLocation().getMember() )
 			);
 
+			// The user didn't set un
```
Seems like something should be following to "un"?

DavideD replied:
Right, while I was writing this comment I had some doubts about the overall behaviour and I've sent the pull request forgetting to remove it.

I will add commits to clean the various left-overs

DavideD commented:
> What would you expect instead if UnwrapValidatedValue is set to false (@UnwrapValidatedValue(false))?
> 
> Then I'd expect @NotNull to be applied to the Optional and @NotNullTypeUse to be applied to the wrapped String.

Should I validate the wrapped element if the validation of the wrapper fails? I'm prone to say no.

gunnarmorling commented on engine-jdk8-tests/src/test/java/org/hibernate/validator/test/internal/engine/valuehandling/OptionalTypeAnnotationConstraintTest.java:
```diff
@@ -0,0 +1,216 @@
+	 * Actual: NotNull violation occurs on container's annotation
+	 */
+	@Test
+	public void not_null_type_on_optional_is_validated_for_empty_value() {
```
This one makes me curious.

The property is annotated with `@UnwrapValidatedValue(false)`, so my personal expectation would be that the type parameter constraint is not applied at all (as it can only apply to the wrapped value, not the wrapper). That's different from the assertion in the test. I think I understand the reasoning between the expectation here, but it seems a bit incosistent to me.

@hferentschik, any thoughts?

DavideD replied:
I think the idea is to be able to validate the wrapper and the wrapped object separately.

By using `@UnwrapValidatedValue(false)`  it saying that the annotation on the property should be used to validate the wrapper.
The validation on the type are used to validate the wrapped object.
In this context, if you set the constraint on the type you always need to unwrap it.

As far as I can see for JDK 8 you don't need the annotation `@UnwrapValidatedValue` but I guess we need to be consistent with the behaviour for JDK 7.

It seems consistent to me, but I don't have much experience with validator.

gunnarmorling commented on engine-jdk8-tests/src/test/java/org/hibernate/validator/test/internal/engine/valuehandling/OptionalTypeAnnotationConstraintTest.java:
```diff
@@ -0,0 +1,216 @@
+		assertCorrectConstraintTypes( constraintViolations, NullOrNotBlank.class );
+	}
+
+	@Test
```
Are the last three actually relevant for the issue at hand?

DavideD replied:
as far as I understand the issue is about to test the ability of validator to validate the wrapper object and the wrapped object separately.

This test is checking that the validation fails because of the constraint on the type and not the one on the wrapper.

There is another one on the same field that checks that when the container is null, the test fails only for the validation of the container `null_or_not_blank_type_on_optional_is_validated_for_null_value`.

hferentschik commented:
>> In particular, what is the expected behaviour for his scenario

> On first thought I'd say both constraints should be applied to the String value

+1

hferentschik commented:
>> What would you expect instead if UnwrapValidatedValue is set to false (@UnwrapValidatedValue(false))?

> Then I'd expect @NotNull to be applied to the Optional and @NotNullTypeUse to be applied to the wrapped String.

Initially I was tempted to say the same, but I see a certain conflict here. I guess one could argue that the type level constraint implies unwrapping, so setting UnwrapValidatedValue to false is either illegal configuration or one ignored the type level constraint and just apply `@NotNull`

DavideD commented:
I have updated the tests and the patch should now work.

I had to change `UnwrapModesTestUnwrapModesTest`, I think it was wrong but I'm not sure, basically the `ValueHandlerConstraintValidator` was receiving as value to validate a `ValueHolder` for both the cases when we use an unwrapper and when we don't. I think that when we use the unwrapper we should validate the value unwrapped value and therfore in that case we should receive an Integer. Because the same validator is used in both test I've changed the type to Object.
This required me to add an `UnwrapValidatedValue ` to solve an exception.

I also addded the tests for method and constructurs.

hferentschik commented:
thanks @DavideD - I am taking a look in a bit ...

DavideD commented:
@hferentschik I think I fixed all the issue

One note:
at the moment the wrapped value gets validated even if the validation of the container fails.

I couldn't do it for methods and constructor and so, for consistency, I did the same for properties. I guess we can open a separate JIRA for this if it is a requirement.

hferentschik commented:
@DavideD, looking good. Unfortunately, I noticed a new bug related bug (see comment in `ConstraintValidatorManager`.

hferentschik commented:
Superseded by #413. 

#413 contains the changes of this pull request + refactoring of `ConstraintValidatorManager` to not modify current `ValueContext`.
